### PR TITLE
Feature/issue 22/implement autowired bean post processor

### PR DESCRIPTION
**Pull request by olehhaliak**
## Checklist before requesting a review
- [x] There are no MR remarks
- [x] Clean and concise code that is easy to read and understand
- [x] Code covered with tests and they are green
- [x] Code API has informative JavaDoc on it
- [x] Rich logging with detailed messages and correct use of log levels

### CODE REVIEW

File: src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  private final BeanDefinitionRegistry definitionRegistry;
+  private final Map<String, Object> rawBeans;
+  private final DependencyResolver dependencyResolver;
+
+  public AutowiredBeanPostProcessor(BeanDefinitionRegistry registry, Map<String, Object> rawBeans) {
+    this.rawBeans = rawBeans;
+    definitionRegistry = registry;
+    dependencyResolver = new DefaultDependencyResolver(definitionRegistry.getAllBeanDefinitions());
+  }
+
```
Comment (vitshch) [inline src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java:28]: Extra empty line

File: src/main/java/com/bringframework/exceptions/BeanInjectionException.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.bringframework.exceptions;
+
```
Comment (marynamelnichuk) [inline src/main/java/com/bringframework/exceptions/BeanInjectionException.java:2]: Please add documentation here, cause for all exceptions classes we have java doc

File: src/test/java/com/bringframework/factory/AutowiredBeanPostProcessorTest.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        "testClassB", secondBean
+    );
+
+    when(dependencyResolver.getCandidateNameOfType(TestClassA.class)).thenReturn("testClassA");
+    autowiredBPP = new AutowiredBeanPostProcessor(beanDefinitionRegistry, rawBeans);
+    injectMockDependencyResolver();
+
+    autowiredBPP.process();
+
+    Assertions.assertSame(testClassABean, secondBean.testClassA);
```
Comment (marynamelnichuk) [inline src/test/java/com/bringframework/factory/AutowiredBeanPostProcessorTest.java]: Use static import here for consistency

File: src/test/java/com/bringframework/resolver/DefaultDependencyResolverTest.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  @Test
+  void candidateResolvedForSameType() {
+    var linkedListBeanDefinition = mock(BeanDefinition.class);
+    doReturn(LinkedList.class).when(linkedListBeanDefinition).getBeanClass();
+
+    var expectedCandidateName = "linkedList";
+    Map<String, BeanDefinition> beanDefinitionMap =
+        Map.of(expectedCandidateName, linkedListBeanDefinition);
+
+    var dependencyResolver = new DefaultDependencyResolver(beanDefinitionMap);
```
Comment (marynamelnichuk) [inline src/test/java/com/bringframework/resolver/DefaultDependencyResolverTest.java]: These snippet is repeated at least 3 times, please create a helper method with this logic.

File: src/test/java/com/bringframework/factory/AutowiredBeanPostProcessorTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class AutowiredBeanPostProcessorTest {
+
+  @Mock
+  DependencyResolver dependencyResolver;
```
Comment (marynamelnichuk) [inline src/test/java/com/bringframework/factory/AutowiredBeanPostProcessorTest.java]: Please add here private modifier and in other lines where it is needed

File: src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+@Slf4j
+public class AutowiredBeanPostProcessor implements BeanPostProcessor {
+  private final BeanDefinitionRegistry definitionRegistry;
+  private final Map<String, Object> rawBeans;
+  private final DependencyResolver dependencyResolver;
+
+  public AutowiredBeanPostProcessor(BeanDefinitionRegistry registry, Map<String, Object> rawBeans) {
+    this.rawBeans = rawBeans;
```
Comment (marynamelnichuk) [inline src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java:24]: Please follow the same style in your code, if you use "this" to set fields in constructor, make the same for all fields.

File: src/test/java/com/bringframework/resolver/DefaultDependencyResolverTest.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  @Test
+  void noCandidateFound_NoSuchBeanExceptionThrown() {
+    var dependencyResolver = new DefaultDependencyResolver(Collections.emptyMap());
+    var desiredCandidateType = Object.class;
+    assertThrows(NoSuchBeanException.class,
+        () -> dependencyResolver.getCandidateNameOfType(desiredCandidateType));
+  }
+
+  @Test
+  void multipleCandidateFound_NoUniqueBeanExceptionThrown() {
```
Comment (marynamelnichuk) [inline src/test/java/com/bringframework/resolver/DefaultDependencyResolverTest.java]: I don't remember whether we agreed on such test methods naming.

**Review by marynamelnichuk — CHANGES_REQUESTED** [review]

File: src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    log.debug("Injecting dependencies for autowired fields complete");
+  }
+
+  private void processBean(String beanName, Object bean) {
+    var beanDefinition = definitionRegistry.getBeanDefinition(beanName);
+    log.debug("Injecting autowired field dependencies for bean {}", beanName);
+    beanDefinition.getAutowiredFieldsMetadata().values()
+        .forEach(field -> populateField(field, bean));
+  }
+
```
Comment (vitshch) [inline src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java:42]: Remove empty line

File: src/test/java/com/bringframework/resolver/DefaultDependencyResolverTest.java (hunk 7 above)
Comment (vitshch) [reply src/test/java/com/bringframework/resolver/DefaultDependencyResolverTest.java]: No, we don't, but I think we have to use **camelCase** in tests naming

**Review by vitshch — CHANGES_REQUESTED** [review]

File: src/main/java/com/bringframework/exceptions/BeanInjectionException.java (hunk 2 above)
Comment (olehhaliak) [reply src/main/java/com/bringframework/exceptions/BeanInjectionException.java:2]: fixed

File: src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java (hunk 6 above)
Comment (olehhaliak) [reply src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java:24]: fixed

File: src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java (hunk 1 above)
Comment (olehhaliak) [reply src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java:28]: fixed

File: src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java (hunk 8 above)
Comment (olehhaliak) [reply src/main/java/com/bringframework/factory/AutowiredBeanPostProcessor.java:42]: fixed

File: src/test/java/com/bringframework/factory/AutowiredBeanPostProcessorTest.java (hunk 3 above)
Comment (olehhaliak) [reply src/test/java/com/bringframework/factory/AutowiredBeanPostProcessorTest.java]: Fixed

File: src/test/java/com/bringframework/resolver/DefaultDependencyResolverTest.java (hunk 4 above)
Comment (olehhaliak) [reply src/test/java/com/bringframework/resolver/DefaultDependencyResolverTest.java]: consolidated this logic in mockBeanDefinitionOfClass() and linkedListBeanDefinitionMap() methods

File: src/test/java/com/bringframework/factory/AutowiredBeanPostProcessorTest.java (hunk 5 above)
Comment (olehhaliak) [reply src/test/java/com/bringframework/factory/AutowiredBeanPostProcessorTest.java]: done

File: src/test/java/com/bringframework/resolver/DefaultDependencyResolverTest.java (hunk 7 above)
Comment (olehhaliak) [reply src/test/java/com/bringframework/resolver/DefaultDependencyResolverTest.java]: renamed in camelCase and added `@DisplayName` instead to preserve readability.

**Review by vitshch — APPROVED** [review]

**Review by MykolaHarasym — APPROVED** [review]

**Review by marynamelnichuk — APPROVED** [review]

**Pull request merged**